File: app/src/main/java/com/popalay/yocard/business/debts/DebtsInteractor.java
package com.popalay.yocard.business.debts;

import com.popalay.yocard.data.models.Debt;
import com.popalay.yocard.data.repositories.DebtsRepository;
import com.popalay.yocard.data.repositories.HolderRepository;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import rx.Completable;
import rx.Observable;
import rx.schedulers.Schedulers;

@Singleton
public class DebtsInteractor {

    private final DebtsRepository debtsRepository;
    private final HolderRepository holderRepository;

    @Inject
    public DebtsInteractor(DebtsRepository debtsRepository, HolderRepository holderRepository) {
        this.debtsRepository = debtsRepository;
        this.holderRepository = holderRepository;
    }

    public Completable save(Debt debt) {
        return debtsRepository.save(debt)
                .subscribeOn(Schedulers.io());
    }

    public Observable<List<Debt>> getDebts() {
        return debtsRepository.getAll()
                .subscribeOn(Schedulers.io());
    }

    public Observable<List<Debt>> getDebtsByHolder(long holderId) {
        return debtsRepository.getAllByHolder(holderId)
                .subscribeOn(Schedulers.io());
    }

    public Completable remove(Debt debt) {
        return debtsRepository.remove(debt)
                .andThen(holderRepository.updateCounts(debt.getHolder()))
                .subscribeOn(Schedulers.io());
    }
}

File: app/src/main/kotlin/com/popalay/cardme/business/interactor/DebtsInteractor.kt
package com.popalay.cardme.business.interactor

import com.popalay.cardme.data.models.Debt
import com.popalay.cardme.data.repositories.DebtRepository
import com.popalay.cardme.data.repositories.HolderRepository
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class DebtsInteractor @Inject constructor(
        private val debtRepository: DebtRepository,
        private val holderRepository: HolderRepository
) {

    fun getAll(): Flowable<List<Debt>> = debtRepository.getAll()
            .subscribeOn(Schedulers.io())

    fun markAsTrash(debt: Debt): Completable = debtRepository.markAsTrash(debt)
            .andThen(holderRepository.removeDebt(debt))
            .subscribeOn(Schedulers.io())

    fun restore(debt: Debt): Completable = holderRepository.addDebt(debt.holder.name, debt)
            .andThen(debtRepository.restore(debt))
            .subscribeOn(Schedulers.io())
}
